lib/Horde/Core/Controller/RequestMapper.php

```diff
@@ -42,34 +140,66 @@ public function getRequestConfiguration(Horde_Injector $injector)
 
         // Push $app onto the registry
         $registry->pushApp($app);
-
         // Application routes are relative only to the application. Let the
         // mapper know where they start.
-        $this->_mapper->prefix = $registry->get('webroot', $app);
-
+        $this->_mapper->prefix = $prefix;
         // Set the application controller directory
         $this->_mapper->directory = $registry->get('fileroot', $app) . '/app/controllers';
 
         // Load application routes.
         $mapper = $this->_mapper;
+        $mapper->environ = array('REQUEST_METHOD' => $request->getMethod());
         include $routeFile;
         if (file_exists($fileroot . '/config/routes.local.php')) {
             include $fileroot . '/config/routes.local.php';
         }
-
         // Match
         // @TODO Cache routes
         $path = $request->getPath();
         if (($pos = strpos($path, '?')) !== false) {
             $path = substr($path, 0, $pos);
         }
+
         $match = $this->_mapper->match($path);
         if (isset($match['controller'])) {
             $config->setControllerName(Horde_String::ucfirst($app) . '_' . Horde_String::ucfirst($match['controller']) . '_Controller');
             $config->setSettingsExporterName($settingsFinder->getSettingsExporterName($config->getControllerName()));
         } else {
             $config->setControllerName('Horde_Core_Controller_NotFound');
         }
+        // TODO: Move to some ControllerAuthHelper and check perms and admin
+        if (!$registry->isAuthenticated()) {
+            $auth = $injector->getInstance('Horde_Core_Factory_Auth')->create();
+
+            // Default behaviour should be to authenticate as older controllers expect it. This should be overrideable
+            if (!isset($match['HordeAuthType'])) {
+                $match['HordeAuthType'] = 'DEFAULT';
+            }
+            // Keep unauthenticated
+            if ($match['HordeAuthType'] == 'NONE') {
+                return $config;
+            }
+            if ($match['HordeAuthType'] == 'DEFAULT') {
+                // Try to authenticate, otherwise redirect to login page
+                // Check for basic auth
+                if (isset($_SERVER['PHP_AUTH_USER']) and isset($_SERVER['PHP_AUTH_PW'])) {
```

Comment:
Wouldn't this always allow Basic Authentication to Horde, no matter if this is intended or not?

Reply:
Yes, that's broken and needs fixing.
Besides, it was kind of a compromise due to BC concerns.
By now, I feel more confident the controller/rampage framework should unconditionally never authenticate but leave that to the controller. It's easier to fix the few controllers we have than live with the conditional. I can provide traits for the typical use cases 

* cookie auth to existing session, redirect to login if invalid
* cookie auth to existing session, error if invalid
* Basic auth
* token header auth (for API use cases)

This would also facilitate switching /rpc/ to controller framework